codegen/src/RenderCommon.hs

```diff
@@ -237,13 +259,116 @@ functionToCpp add_type_initials prefix fn = [st|
       if add_type_initials
       then "_" <> (mconcat $ flip map parameters' $ \p -> parsableToInitial (ptype p))
       else ""
+    pointer :: Text
+    pointer =
+      if is_managed
+      then "ForeignPtr"
+      else "Ptr"
     types_list :: [Text]
     types_list = flip map parameters' $ \p ->
       if isCType (ptype p)
       then [st|#{parsableToHsType (ptype p)}|]
-      else [st|Ptr #{parsableToHsType (ptype p)}|]
+      else [st|#{pointer} #{parsableToHsType (ptype p)}|]
+    types :: Text
+    types = T.intercalate "\n  -> " $ types_list ++ [[st|IO (#{ret_hstype})|]]
+    ret_type :: Text
+    ret_type =
+      if isCType (retType fn)
+      then [st|#{parsableToCppType (retType fn)}|]
+      else [st|#{parsableToCppType (retType fn)}*|]
+    ret_hstype :: Text
+    ret_hstype =
+      if isCType (retType fn)
+      then [st|#{parsableToHsType (retType fn)}|]
+      else [st|#{pointer} #{parsableToHsType (retType fn)}|]
+    ret_wrapper :: Text
+    ret_wrapper =
+      if isCType (retType fn)
+      then ""
+      else [st|new #{parsableToCppType (retType fn)}|]
+    call_return :: Text
+    call_return =
+      case (retType fn) of
+        CType CVoid -> ""
+        _ -> "return"
+
+methodToCpp :: PC.CppClass -> Bool -> Bool -> Bool -> String -> String -> Function -> Text
```

Comment:
functionToCpp generates codes for top-level-functions.
methodToCpp generates codes for c++-objects.
I want to refactor these functions because there are many places where the functions are duplicated